### ISSUE TITLE
dynamically loaded code snippets via HTMX missing "hljs" class on subsequent element swaps

### ISSUE THREAD

**Issue by pwntr** [highlightjs/highlight.js#4163]
**Describe the issue/behavior that seems buggy**
I use HTMX to load code snippets from my go backend, dynamically on clicking a button. The backend templates the desired code inside the `<pre><code>` tags and returns it to the requesting browser. The buggy behavior is that only on the first swapping-in of the templated element the "hljs" class is applied to the element that is to be highlighted. If I load a new code snippet without a full page refresh, the "hljs" class is now missing from the classes attribute of that element. The code lines themselves are still highlighted, but the surrounding style container and padding provided by the "hljs" class are missing.

The clarify, I swap the "inner HTML" in HTMX terms, meaning I fully replace all the templated data on each swap. The server returns the data as presented below (with the code to style obviously present).

To workaround this class omission, I need to manually add the "hljs" class to my code tag's classes attribute inside the template. After that, it works consistently on the first and on any subsequent element swaps, with the "hljs" class present.

I tried this with multiple styles incl. the default one.

**Sample Code or Instructions to Reproduce**

Template code (only working when swapping this in for the first time):
```HTML
<pre><code class="language-yaml" id="code-data">{{ . }}</code></pre>
<script>hljs.highlightElement(document.getElementById('code-data'));</script>
```

Template code (workaround, with manually added hljs class):
```HTML
<pre><code class="language-yaml hljs" id="code-data">{{ . }}</code></pre>
<script>hljs.highlightElement(document.getElementById('code-data'));</script>
```

**Expected behavior**
given that I specify the the element's ID to be styled, and having verified that the script fires correctly on each swap-in of the templated data listed above, I would expect the "hljs" class to be added correctly also on subsequent swaps, since the entire code element is always replaced.

**Additional context**
latest versions of highlight.js and htmx used, tested in chrome and firefox

**Comment by joshgoebel** [comment]
If you check the `highlighted` dataset attribute in-between do you find it set to `true`?  We will refuse to highlight in that case - you'd have to first clear status that from the dataset before attempting to highlight the same element more than once.

Otherwise this is very, very simple code in our library: https://github.com/highlightjs/highlight.js/blob/main/src/highlight.js#L727

If you can create an isolated JSFiddle with just our library (no HTMX, etc) reproducing the issue, we'd be happy to have another look.  But if it only happens with HTMX, perhaps it's a HTMX bug, not us.

**Comment by joshgoebel** [comment]
Also check your console for warnings, etc.

**Comment by pwntr** [comment]
Thank you for those hints and your help, Josh! I got no error or other messages as indicated by the source you linked, but I went down the rabbit hole and tried to verify my observations with a [jsfiddle](https://jsfiddle.net/2ryxdsz4/23/). Turns out, with vanilla JS and no templated returning of script tags on my side, it works just fine when replacing the entire container. No need to remove the attribute or pre-add the `hljs` class if I remove it all. However, styling state was also indeed honored correctly limited to that dataset attribute that I could remove when loading in content dynamically into a static container. So with that understanding I reworked my approach entirely (see below).

I will close this issue here and leave a few explanations on what I did to properly solve this for my specific usage with HTMX, maybe another wanderer might find this to be useful in the future.

I revamped my entire templated script inclusion, and actually got rid of the templated code+script tag approach as shown above all together. Now I just return the naked code and apply the styling in a static container that gets re-styled after the HTMX swap-in, and reset (the dataset attribute) before swap-in, like this:

```HTML
<pre><code class="language-yaml" id="code-container" hx-on--before-swap="this.removeAttribute('data-highlighted');" hx-on--after-swap="hljs.highlightElement(this);"></code></pre>
```

The button that triggers this now looks like this, to allow for a loading message to be shown inside the `<pre><code>` tags while the data is being fetched from the backend:

```HTML
<a href="#" hx-get="/api/code/{{ .Id }}/profile" hx-target="#code-container" hx-on--before-request="event.detail.target.classList.remove('hljs'); event.detail.target.innerHTML='loading profile...';" class="btn btn-sm btn-outline-secondary" data-bs-toggle="modal" data-bs-target="#code-modal">View</a>
```

I can now safely swap different code snippets in and out dynamically, same modal container, no problem.

Thanks a again and have a nice weekend :relieved:!
